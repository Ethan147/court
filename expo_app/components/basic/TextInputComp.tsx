import React, { useState } from "react";
import {
  NativeSyntheticEvent,
  TextInputFocusEventData,
  TouchableOpacity,
  View,
  StyleProp,
  ViewStyle,
  Text,
} from "react-native";
import { TextInput } from "react-native-paper";
import { MaterialIcons } from "@expo/vector-icons";
import colors from "../../utils/colors";
import theme from "../../utils/theme";

const customTheme = {
  ...theme,
  colors: {
    ...theme.colors,
    primary: colors.primary,
    placeholder: colors.text,
  },
};

const TextInputComp = ({
  label,
  value,
  onChangeText,
  onBlur,
  error,
  secureTextEntry,
  passStyles,
  hintMessage,
}: {
  label?: string;
  value?: string;
  onChangeText?: (text: string) => void;
  onBlur?: (e: NativeSyntheticEvent<TextInputFocusEventData>) => void;
  error?: boolean;
  secureTextEntry?: boolean;
  passStyles?: {
    textInputCompOuterView?: StyleProp<ViewStyle>;
    textInputCompTextContainer?: StyleProp<ViewStyle>;
    textInputCompIconContainer?: StyleProp<ViewStyle>;
    textInputCompViewStyle?: StyleProp<ViewStyle>;
    textInputCompText?: StyleProp<ViewStyle>;
    textInputCompHintMessage?: StyleProp<ViewStyle>;
  };
  hintMessage?: string;
}) => {
  const styles = {
    textInputCompOuterView: passStyles?.textInputCompOuterView || {},
    textInputCompTextContainer: passStyles?.textInputCompTextContainer || {},
    textInputCompIconContainer: passStyles?.textInputCompIconContainer || {},
    textInputCompViewStyle: passStyles?.textInputCompViewStyle || {},
    textInputCompText: passStyles?.textInputCompText || {},
    textInputCompHintMessage: passStyles?.textInputCompHintMessage || {},
  };

  const [showPassword, setShowPassword] = useState(false);

  return (
    <View style={styles.textInputCompOuterView}>
      <View style={styles.textInputCompTextContainer}>
        <TextInput
          label={label}
          value={value}
          onChangeText={onChangeText}
          onBlur={onBlur}
          error={error}
          autoComplete={undefined}
          style={styles.textInputCompText}
          theme={customTheme}
          secureTextEntry={secureTextEntry && !showPassword}
        />
        {secureTextEntry && (
          <TouchableOpacity
            onPress={() => setShowPassword(!showPassword)}
            style={styles.textInputCompIconContainer}
          >
            <MaterialIcons
              name={showPassword ? "visibility" : "visibility-off"}
              size={theme.font.size.large}
              color={"gray"}
            />
          </TouchableOpacity>
        )}
      </View>
      {hintMessage && (
        <Text style={styles.textInputCompHintMessage}>{hintMessage}</Text>
      )}
    </View>
  );
};

export default TextInputComp;
