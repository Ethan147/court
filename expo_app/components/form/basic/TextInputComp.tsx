import React, { useState } from "react";
import {
  NativeSyntheticEvent,
  TextInputFocusEventData,
  TouchableOpacity,
  View,
  StyleProp,
  ViewStyle,
} from "react-native";
import { TextInput } from "react-native-paper";
import { MaterialIcons } from "@expo/vector-icons";
import colors from "../../../utils/colors";
import theme from "../../../utils/theme";

const customTheme = {
  ...theme,
  colors: {
    ...theme.colors,
    primary: colors.primary,
    placeholder: colors.text,
  },
};

const TextInputComp = ({
  label,
  value,
  onChangeText,
  onBlur,
  error,
  secureTextEntry,
  passStyles,
}: {
  label?: string;
  value?: string;
  onChangeText?: (text: string) => void;
  onBlur?: (e: NativeSyntheticEvent<TextInputFocusEventData>) => void;
  error?: boolean;
  secureTextEntry?: boolean;
  passStyles?: {
    textInputCompContainer?: StyleProp<ViewStyle>;
    textInputCompIconContainer?: StyleProp<ViewStyle>;
    textInputCompViewStyle?: StyleProp<ViewStyle>;
    textInputCompText?: StyleProp<ViewStyle>;
  };
}) => {
  const styles = {
    textInputCompContainer: passStyles?.textInputCompContainer || {},
    textInputCompIconContainer: passStyles?.textInputCompIconContainer || {},
    textInputCompViewStyle: passStyles?.textInputCompViewStyle || {},
    textInputCompText: passStyles?.textInputCompText || {},
  };

  const [showPassword, setShowPassword] = useState(false);

  return (
    <View style={styles.textInputCompViewStyle}>
      <View style={styles.textInputCompContainer}>
        <TextInput
          label={label}
          value={value}
          onChangeText={onChangeText}
          onBlur={onBlur}
          error={error}
          autoComplete={undefined}
          style={styles.textInputCompText}
          theme={customTheme}
          secureTextEntry={secureTextEntry && !showPassword}
        />
        {secureTextEntry && (
          <TouchableOpacity
            onPress={() => setShowPassword(!showPassword)}
            style={styles.textInputCompIconContainer}
          >
            <MaterialIcons
              name={showPassword ? "visibility" : "visibility-off"}
              size={theme.font.size.large}
              color={"gray"}
            />
          </TouchableOpacity>
        )}
      </View>
    </View>
  );
};

export default TextInputComp;
