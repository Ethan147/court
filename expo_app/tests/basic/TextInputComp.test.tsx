import React from "react";
import { render, fireEvent } from "@testing-library/react-native";

import TextInputComp from "../../components/basic/TextInputComp";

describe("<TextInputComp />", () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it("renders without crashing", () => {
    const { getByTestId } = render(<TextInputComp />);
    expect(getByTestId("textInputComp")).toBeDefined();
  });

  // targeting the label exactly may not be doable with jest and react-native-paper's TextInput
  it("renders without crashing when label prop is passed", () => {
    const { getByTestId } = render(<TextInputComp label="Test Label" />);
    expect(getByTestId("textInputComp")).toBeDefined();
  });

  it("displays the correct initial value", () => {
    const { getByDisplayValue } = render(
      <TextInputComp value="Initial value" />
    );
    expect(getByDisplayValue("Initial value")).toBeDefined();
  });

  it("calls the onChangeText function when text is entered", () => {
    const onChangeTextMock = jest.fn();
    const { getByTestId } = render(
      <TextInputComp onChangeText={onChangeTextMock} />
    );
    const input = getByTestId("textInputComp");

    fireEvent.changeText(input, "new text");
    expect(onChangeTextMock).toHaveBeenCalledWith("new text");
  });

  it("calls the onBlur function when focus is lost", () => {
    const onBlurMock = jest.fn();
    const { getByTestId } = render(<TextInputComp onBlur={onBlurMock} />);
    const input = getByTestId("textInputComp");

    input.props.onBlur();
    expect(onBlurMock).toHaveBeenCalled();
  });

  it("updates on change", () => {
    const { getByTestId } = render(<TextInputComp />);
    const input = getByTestId("textInputComp");

    fireEvent.changeText(input, "test value");
    expect(input.props.value).toBe("test value");
  });

  it("calls onChange callback with changed text", () => {
    const onChange = jest.fn();
    const { getByTestId } = render(<TextInputComp onChangeText={onChange} />);
    const input = getByTestId("textInputComp");

    fireEvent.changeText(input, "test value");
    expect(onChange).toHaveBeenCalledWith("test value");
  });

  it("secureTextEntry prop hides the text until 'show password' button is clicked", () => {
    const { getByTestId } = render(<TextInputComp secureTextEntry />);
    const input = getByTestId("textInputComp");
    const showPasswordButton = getByTestId("togglePasswordVisibility");

    expect(input.props.secureTextEntry).toBe(true);

    // After clicking the 'show password' button, the text should be visible
    fireEvent.press(showPasswordButton);
  });

  it("secureTextEntry prop hides the text again when 'show password' button is clicked twice", () => {
    let secureTextEntry = true;
    const { getByTestId, rerender } = render(
      <TextInputComp secureTextEntry={secureTextEntry} />
    );
    const input = getByTestId("textInputComp");
    const showPasswordButton = getByTestId("togglePasswordVisibility");

    // Initially, the text should be hidden
    expect(input.props.secureTextEntry).toBe(true);

    // After clicking the 'show password' button once, the text should be visible
    secureTextEntry = false;
    rerender(<TextInputComp secureTextEntry={secureTextEntry} />);
    expect(input.props.secureTextEntry).toBe(false);

    // After clicking the 'show password' button again, the text should be hidden again
    secureTextEntry = true;
    rerender(<TextInputComp secureTextEntry={secureTextEntry} />);
    expect(input.props.secureTextEntry).toBe(true);
  });

  it("snapshot", () => {
    const tree = render(<TextInputComp />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
