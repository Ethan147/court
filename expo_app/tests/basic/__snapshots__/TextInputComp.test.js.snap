// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInputComp /> snapshot 1`] = `
<View
  style={{}}
>
  <View
    style={{}}
  >
    <View
      style={
        [
          {
            "backgroundColor": "rgba(231, 224, 236, 1)",
            "borderTopLeftRadius": 4,
            "borderTopRightRadius": 4,
          },
          {},
        ]
      }
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "rgba(28, 27, 31, 1)",
            "bottom": 0,
            "height": 1,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "transform": [
              {
                "scaleY": 0.5,
              },
            ],
            "zIndex": 1,
          }
        }
        testID="text-input-underline"
      />
      <View
        style={
          [
            {
              "paddingBottom": 0,
              "paddingTop": 0,
            },
            {
              "minHeight": 56,
            },
          ]
        }
      >
        <TextInput
          cursorColor="#40916C"
          editable={true}
          maxFontSizeMultiplier={1.5}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholderTextColor="rgba(73, 69, 79, 1)"
          selectionColor="#40916C"
          style={
            [
              {
                "margin": 0,
              },
              {
                "height": 56,
              },
              {
                "paddingBottom": 0,
                "paddingTop": 0,
              },
              {
                "color": "rgba(73, 69, 79, 1)",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "letterSpacing": 0.15,
                "lineHeight": 19.2,
                "minWidth": 65,
                "paddingLeft": 16,
                "paddingRight": 16,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
              false,
              [
                {},
              ],
              undefined,
            ]
          }
          testID="textInputComp"
          underlineColorAndroid="transparent"
        />
      </View>
    </View>
  </View>
</View>
`;
